mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java

```diff
@@ -278,26 +278,26 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
 
     private Mono<Task.Result> reIndexMessages(Flux<Either<Failure, ReIndexingEntry>> entriesToIndex, RunningOptions runningOptions, ReprocessingContext reprocessingContext) {
         return ReactorUtils.Throttler.<Either<Failure, ReIndexingEntry>, Task.Result>forOperation(
-                entry -> reIndexMessage(entry, reprocessingContext))
+                entry -> reIndex(entry, reprocessingContext))
             .window(runningOptions.getMessagesPerSecond(), Duration.ofSeconds(1))
             .throttle(entriesToIndex)
             .reduce(Task::combine)
             .switchIfEmpty(Mono.just(Result.COMPLETED));
     }
 
-    private Mono<Task.Result> reIndexMessage(Either<Failure, ReIndexingEntry> entryOrFailure, ReprocessingContext reprocessingContext) {
-        return toMono(entryOrFailure.right().map(this::index))
-            .map(this::flapMapRight)
-            .map(either -> recordIndexingResult(reprocessingContext, either));
+    private Mono<Task.Result> reIndex(Either<Failure, ReIndexingEntry> entryOrFailure, ReprocessingContext reprocessingContext) {
+        return entryOrFailure.map(this::index)
+            .fold(failure -> {
+                failure.recordFailure(reprocessingContext);
+                return Mono.just(Result.PARTIAL);
+            }, result -> result.map(either -> recordIndexingResult(either, reprocessingContext)));
     }
 
-    private Result recordIndexingResult(ReprocessingContext reprocessingContext, Either<Failure, Result> either) {
-        return either.either(
-            failure -> {
-                failure.recordFailure(reprocessingContext);
-                return Result.PARTIAL;
-            },
-            result -> result.onComplete(reprocessingContext::recordSuccess));
+    private Result recordIndexingResult(Either<Failure, Result> either, ReprocessingContext reprocessingContext) {
+        return either.fold(failure -> {
+            failure.recordFailure(reprocessingContext);
+            return Result.PARTIAL;
```

Comment:
This looks duplicated with reIndex(Either<Failure, ReIndexingEntry> entryOrFailure, ReprocessingContext reprocessingContext) (l 290 -> 293)